## List of contents
1. Libraries
2. Data Sets
3. Module Exercises
4. Task 4.5

## 1. Libraries

In [3]:
# Imported libraries

In [5]:
import pandas as pd
import numpy as np
import os

## 2. Data Sets

In [8]:
# Creating folder part

In [10]:
path=r'/Users/konstant/Documents/Achievement 4. Instacart Basket Analysis'

In [12]:
path

'/Users/konstant/Documents/Achievement 4. Instacart Basket Analysis'

In [14]:
# Importing Data Sets - “products.csv” data set from your “Original Data” folder as df_prods

In [16]:
df_prods= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [18]:
# Importing Data Sets - “orders_wrangled.csv” data set from your “Prepared Data” folder as df_ords.

In [20]:
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [22]:
# (Note that the index column is set to 0 otherwise a 'unnamed: 0' column is also imported)

In [24]:
df_prods.shape

(49693, 5)

## 3. Module Exercise

In [54]:
df_ords.describe()

,order_id,user_id,order_number,orders_day_of_week,hour_of_purchase,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


In [56]:
# creating a data frame

In [58]:
df_test=pd.DataFrame()

In [60]:
df_test['mix']=['a','b',1,True]

In [62]:
df_test.head()

,mix
0,a
1,b
2,1
3,True


In [64]:
# Check for mixed data type

In [66]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [68]:
#Fix to change the data type

In [70]:
df_test['mix'] = df_test['mix'].astype('str')

In [72]:
# Finding missing values in the df_prods data frame

In [74]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [76]:
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [78]:
df_prods_clean

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
1,2,All-Seasons Salt,104,13,9.3
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1,10.5
4,5,Green Chile Anytime Sauce,5,13,4.3
...,...,...,...,...,...
49688,49684,"Vodka, Triple Distilled, Twist of Vanilla",124,5,5.3
49689,49685,En Croute Roast Hazelnut Cranberry,42,1,3.1
49690,49686,Artisan Baguette,112,3,7.8
49691,49687,Smartblend Healthy Metabolism Dry Cat Food,41,8,4.7


In [80]:
df_prods.shape

(49693, 5)

In [82]:
df_prods_clean.shape

(49677, 5)

In [84]:
# checking for duplicates

In [86]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [88]:
df_dups

,product_id,product_name,aisle_id,department_id,prices
462,462,Fiber 4g Gummy Dietary Supplement,70,11,4.8
18459,18458,Ranger IPA,27,5,9.2
26810,26808,Black House Coffee Roasty Stout Beer,27,5,13.4
35309,35306,Gluten Free Organic Peanut Butter & Chocolate ...,121,14,6.8
35495,35491,Adore Forever Body Wash,127,11,9.9


In [90]:
# Removing duplicates

In [92]:
df_prods_clean.shape

(49677, 5)

In [94]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [96]:
df_prods_clean_no_dups.shape

(49672, 5)

In [98]:
# exporting the cleaned product data set. 

In [100]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 4. Task 4.5

In [103]:
# Turning off the scientific notation for floats

In [105]:
pd.set_option('display.float_format', '{:.2f}'.format)

# 1. Consistency check on df_prods

In [108]:
df_prods.describe()

,product_id,aisle_id,department_id,prices
count,49693.00,49693.00,49693.00,49693.00
mean,24844.35,67.77,11.73,9.99
std,14343.72,38.32,5.85,453.52
min,1.00,1.00,1.00,1.00
25%,12423.00,35.00,7.00,4.10
50%,24845.00,69.00,13.00,7.10
75%,37265.00,100.00,17.00,11.20
max,49688.00,134.00,21.00,99999.00


## - We cant read much on the statistical analysis of the product_id, aisle_id and departmen_id as these are string values. The count matches all 3 columns, which is correct.
## - On the price, the max value looks like an outlier as the mean is only 9.99 and the max is 99999.00, which is an amomaly

# 2. Consistency check on df_ords

In [112]:
df_ords.describe()

,order_id,user_id,order_number,orders_day_of_week,hour_of_purchase,days_since_prior_order
count,3421083.00,3421083.00,3421083.00,3421083.00,3421083.00,3214874.00
mean,1710542.00,102978.21,17.15,2.78,13.45,11.11
std,987581.74,59533.72,17.73,2.05,4.23,9.21
min,1.00,1.00,1.00,0.00,0.00,0.00
25%,855271.50,51394.00,5.00,1.00,10.00,4.00
50%,1710542.00,102689.00,11.00,3.00,13.00,7.00
75%,2565812.50,154385.00,23.00,5.00,16.00,15.00
max,3421083.00,206209.00,100.00,6.00,23.00,30.00


## - we cannot read much from the statistical analysis on the order_id, user_id or order_number as these are string values. The count matches, which is correct.
## - orders_day_of_week: everything seems normal. The min starts at 0 and the max end at 6, indicating 7 days of the week. No negative values 
## - hour_of_purchase: everything seems normal here. The min starts at 0, and the max ends at 23. No negative, which is also good.
## - days_since_prior_order: everything seems normal. No outliers and no negatives.

# 3. Checking for mixed data types. 

In [116]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# 4. Note that the column: 'days_since_prior_order' could not be changed to numbers (int64) as it contained Na values

# 5. Checking for missing values

In [120]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_purchase               0
days_since_prior_order    206209
dtype: int64

## from the output of this code, we can see that in the column 'days_since_prior_order' there are 206,209 cell observations with missing data.
## This could mean that these are new orders, but since this column also contains the numeric value of 0, further investigation is needed
## It could refer to cancelled or returned orders, where the days_since_prior_order would be blank, and since we have no order status in this data set, further investigation is needed.

# 6. Addressing the missing values.

In [124]:
# Checking what the missing vlaue looks like. 

In [126]:
df_missing_values=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [128]:
df_missing_values

,order_id,user_id,order_number,orders_day_of_week,hour_of_purchase,days_since_prior_order
0,2539329,1,1,2,8,NaN
11,2168274,2,1,2,11,NaN
26,1374495,3,1,1,14,NaN
39,3343014,4,1,6,11,NaN
45,2717275,5,1,3,12,NaN
...,...,...,...,...,...,...
3420930,969311,206205,1,4,12,NaN
3420934,3189322,206206,1,3,18,NaN
3421002,2166133,206207,1,6,19,NaN
3421019,2227043,206208,1,1,15,NaN


## Due to the high number of missing vlaues, its best to do nothing as these missing vlaues could have a significant meaning.

In [131]:
df_ords

,order_id,user_id,order_number,orders_day_of_week,hour_of_purchase,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.00
2,473747,1,3,3,12,21.00
3,2254736,1,4,4,7,29.00
4,431534,1,5,4,15,28.00
...,...,...,...,...,...,...
3421078,2266710,206209,10,5,18,29.00
3421079,1854736,206209,11,4,10,30.00
3421080,626363,206209,12,1,12,18.00
3421081,2977660,206209,13,1,12,7.00


In [133]:
# rechecking for blank values 

In [135]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_purchase               0
days_since_prior_order    206209
dtype: int64

# 7. Checking for duplicate values

In [138]:
df_dups = df_ords[df_ords.duplicated()]

In [140]:
df_dups

,order_id,user_id,order_number,orders_day_of_week,hour_of_purchase,days_since_prior_order


## No duplicated values were found in the data frame df_ords.

# 8. No method was neccesary for dealing with the duplicates, but one of the most effective methods to deal with them would be to delete them.

# 9. Exporting the final cleaned data. 

In [145]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [26]:
df_ords.shape

(3421083, 6)